pkg/filters/proxy/wspool.go

```diff
@@ -0,0 +1,233 @@
+		resp, _ = httpprot.NewResponse(nil)
+	}
+
+	resp.SetStatusCode(statusCode)
+	ctx.SetOutputResponse(resp)
+}
+
+func (sp *WebSocketServerPool) dialServer(svr *Server, req *httpprot.Request) (*websocket.Conn, error) {
+	origin := req.HTTPHeader().Get("Origin")
+	if len(origin) == 0 {
+		origin = sp.proxy.spec.DefaultOrigin
+	}
+
+	u := *req.URL()
+	u1, err := url.ParseRequestURI(svr.URL)
+	if err != nil {
+		return nil, err
+	}
+	u.Host = u1.Host
+	switch u1.Scheme {
+	case "ws", "wss":
+		u.Scheme = u1.Scheme
+		break
+	case "http":
+		u.Scheme = "ws"
+	case "https":
+		u.Scheme = "wss"
+	default:
+		return nil, fmt.Errorf("invalid scheme %s", u1.Scheme)
+	}
+
+	config, err := websocket.NewConfig(u.String(), origin)
+	if err != nil {
+		return nil, err
+	}
+
+	// copies headers from req to the dialer and forward them to the
+	// destination.
+	//
+	// According to https://docs.oracle.com/en-us/iaas/Content/Balance/Reference/httpheaders.htm
+	// For load balancer, we add following key-value pairs to headers
+	// X-Forwarded-For: <original_client>, <proxy1>, <proxy2>
+	// X-Forwarded-Host: www.example.com:8080
+	// X-Forwarded-Proto: https
+	//
+	// The websocket library discards some of the headers, so we just clone
+	// the original headers and add the above headers.
+	config.Header = req.HTTPHeader().Clone()
+
+	const xForwardedFor = "X-Forwarded-For"
+	xff := req.HTTPHeader().Get(xForwardedFor)
+	if clientIP, _, err := net.SplitHostPort(req.RemoteAddr); err == nil {
+		if xff == "" {
+			config.Header.Set(xForwardedFor, clientIP)
+		} else {
+			config.Header.Set(xForwardedFor, fmt.Sprintf("%s, %s", xff, clientIP))
+		}
+	}
+
+	const xForwardedHost = "X-Forwarded-Host"
+	xfh := req.HTTPHeader().Get(xForwardedHost)
+	if xfh == "" && req.Host() != "" {
+		config.Header.Set(xForwardedHost, req.Host())
+	}
+
+	const xForwardedProto = "X-Forwarded-Proto"
+	config.Header.Set(xForwardedProto, "http")
+	if req.TLS != nil {
+		config.Header.Set(xForwardedProto, "https")
+	}
+
+	return websocket.DialConfig(config)
+}
+
+func (sp *WebSocketServerPool) handle(ctx *context.Context) (result string) {
+	req := ctx.GetInputRequest().(*httpprot.Request)
+	svr := sp.LoadBalancer().ChooseServer(req)
+
+	metric := &httpstat.Metric{}
+	startTime := fasttime.Now()
+	defer func() {
+		metric.Duration = fasttime.Since(startTime)
+		sp.httpStat.Stat(metric)
+	}()
+
+	// if there's no available server.
+	if svr == nil {
+		logger.Debugf("%s: no available server", sp.name)
+		sp.buildFailureResponse(ctx, http.StatusServiceUnavailable)
+		metric.StatusCode = http.StatusServiceUnavailable
+		return resultInternalError
+	}
+
+	stdw, _ := ctx.GetData("HTTP_RESPONSE_WRITER").(http.ResponseWriter)
+	if stdw == nil {
+		logger.Debugf("%s: cannot get response writer from context", sp.name)
+		sp.buildFailureResponse(ctx, http.StatusInternalServerError)
+		metric.StatusCode = http.StatusInternalServerError
+		return resultInternalError
+	}
+
+	wssvr := websocket.Server{}
```

Comment:
propose to use `websocket.Handler` here. `websocket.Server` may looks heavy here. 
```
handler := websocket.Handler(func(clntConn *websocket.Conn) {
		// dial to the server
		svrConn, err := sp.dialServer(svr, req)
                ...
})

handler.ServeHTTP(stdw, req.Request)
```

Reply:
actually, `websocket.Handler` creates a `websocket.Server`, so it is more heavy.
and we have to use `websocket.Server` because we need to use a nil handshaker.